### PR TITLE
[Modules] Updated Blob Services for Storage Account Module to latest API and added new parameters

### LINKED ISSUE #2994: [Feature Request]: storageaccount/blobservices: expose field to control container softdelete
### Description

the template today https://github.com/Azure/ResourceModules/blob/main/modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep#L110 only exposes `deleteRetentionPolicy` and not `containerDeleteRetentionPolicy`
Feature is to expose these attributes as documented on: https://learn.microsoft.com/en-us/azure/storage/blobs/soft-delete-container-enable?tabs=template

### PR DESCRIPTION
**Pull request by ahmadabdalla**
# Description

Closes #2994 

| Pipeline |
| - |
|[Storage - StorageAccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: modules/Microsoft.Storage/storageAccounts/blobServices/containers/immutabilityPolicies/deploy.bicep [hunk 1]
```diff
@@ -11,6 +11,9 @@ param immutabilityPeriodSinceCreationInDays int = 365
 @description('Optional. This property can only be changed for unlocked time-based retention policies. When enabled, new blocks can be written to an append blob while maintaining immutability protection and compliance. Only new blocks can be added and any existing blocks cannot be modified or deleted. This property cannot be changed with ExtendImmutabilityPolicy API.')
 param allowProtectedAppendWrites bool = true
 
+@description('Optional. This property can only be changed for unlocked time-based retention policies. When enabled, new blocks can be written to both "Append and Bock Blobs" while maintaining immutability protection and compliance. Only new blocks can be added and any existing blocks cannot be modified or deleted. This property cannot be changed with ExtendImmutabilityPolicy API. The "allowProtectedAppendWrites" and "allowProtectedAppendWritesAll" properties are mutually exclusive.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/containers/immutabilityPolicies/deploy.bicep:14]:
```suggestion
@description('Optional. This property can only be changed for unlocked time-based retention policies. When enabled, new blocks can be written to both "Append and Block Blobs" while maintaining immutability protection and compliance. Only new blocks can be added and any existing blocks cannot be modified or deleted. This property cannot be changed with ExtendImmutabilityPolicy API. The "allowProtectedAppendWrites" and "allowProtectedAppendWritesAll" properties are mutually exclusive.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 2]
```diff
@@ -2,14 +2,52 @@
...
-param deleteRetentionPolicy bool = true
+@description('Optional. Automatic Snapshot is enabled if set to true.')
+param automaticSnapshotPolicyEnabled bool = false
+
+@description('Optional. The blob service properties for change feed events. Indicates whether change feed event logging is enabled for the Blob service.')
+param changeFeedEnabled bool = true
+
+@minValue(0)
+@maxValue(146000)
+@description('Optional. Indicates whether change feed event logging is enabled for the Blob service. Indicates the duration of changeFeed retention in days. Minimum value is 1 day and maximum value is 146000 days (400 years). A "0" value indicates an infinite retention of the change feed.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:13]:
```suggestion
@description('Optional. Indicates whether change feed event logging is enabled for the Blob service. Indicates the duration of changeFeed retention in days. A "0" value indicates an infinite retention of the change feed.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 3]
```diff
@@ -2,14 +2,52 @@
...
+param changeFeedRetentionInDays int = 7
+
+@description('Optional. The blob service properties for container soft delete. Indicates whether DeleteRetentionPolicy is enabled.')
+param containerDeleteRetentionPolicyEnabled bool = true
+
+@minValue(1)
+@maxValue(365)
+@description('Optional. Indicates the number of days that the deleted item should be retained. The minimum specified value can be 1 and the maximum value can be 365.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:21]:
```suggestion
@description('Optional. Indicates the number of days that the deleted item should be retained.')

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 4]
```diff
@@ -2,14 +2,52 @@
...
+
+@description('Optional. The blob service properties for blob soft delete. Indicates whether DeleteRetentionPolicy is enabled for the Blob service.')
+param deleteRetentionPolicyEnabled bool = true
 
 @description('Optional. Indicates the number of days that the deleted blob should be retained. The minimum specified value can be 1 and the maximum value can be 365.')
 param deleteRetentionPolicyDays int = 7
 
-@description('Optional. Automatic Snapshot is enabled if set to true.')
-param automaticSnapshotPolicyEnabled bool = false
+@description('Optional. Use versioning to automatically maintain previous versions of your blobs. Versioning is enabled if set to true.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:39]:
```suggestion
@description('Optional. Use versioning to automatically maintain previous versions of your blobs.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 5]
```diff
@@ -2,14 +2,52 @@
...
+param isVersioningEnabled bool = true
+
+@description('Optional. The blob service property to configure last access time based tracking policy. When set to true last access time based tracking is enabled.')
+param lastAccessTimeTrackingPolicyEnable bool = false
+
+@description('Optional. The blob service properties for blob restore policy. Blob restore is enabled if set to true. If point-in-time restore is enabled, then versioning, change feed, and blob soft delete must also be enabled.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:45]:
```suggestion
@description('Optional. The blob service properties for blob restore policy. If point-in-time restore is enabled, then versioning, change feed, and blob soft delete must also be enabled.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 6]
```diff
@@ -2,14 +2,52 @@
...
+@description('Optional. This property when set to true allows deletion of the soft deleted blob versions and snapshots. This property cannot be used blob restore policy. This property only applies to blob service and does not apply to containers or file share.')
+param containerDeleteRetentionPolicyAllowPermanentDelete bool = false
+
+@description('Optional. Specifies CORS rules for the Blob service. You can include up to five CorsRule elements in the request. If no CorsRule elements are included in the request body, all CORS rules will be deleted, and CORS will be disabled for the Blob service.')
+param corsRules array = []
+
+@description('Optional. Indicates the default version to use for requests to the Blob service if an incoming request\'s version is not specified. Possible values include version 2008-10-27 and all more recent versions.')
+param defaultServiceVersion string = ''
+
+@description('Optional. The blob service properties for blob soft delete. Indicates whether DeleteRetentionPolicy is enabled for the Blob service.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:33]:
```suggestion
@description('Optional. The blob service properties for blob soft delete.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 7]
```diff
@@ -2,14 +2,52 @@
...
+param deleteRetentionPolicyEnabled bool = true
 
 @description('Optional. Indicates the number of days that the deleted blob should be retained. The minimum specified value can be 1 and the maximum value can be 365.')
 param deleteRetentionPolicyDays int = 7
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:37]: We should remove the limits from the description and use the Metadata limitations instead

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep [hunk 8]
```diff
@@ -2,14 +2,52 @@
...
+param isVersioningEnabled bool = true
+
+@description('Optional. The blob service property to configure last access time based tracking policy. When set to true last access time based tracking is enabled.')
+param lastAccessTimeTrackingPolicyEnable bool = false
+
+@description('Optional. The blob service properties for blob restore policy. Blob restore is enabled if set to true. If point-in-time restore is enabled, then versioning, change feed, and blob soft delete must also be enabled.')
+param restorePolicyEnabled bool = true
+
+@minValue(1)
+@description('Optional. how long this blob can be restored. It should be great than zero and less than DeleteRetentionPolicy days.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:49]:
```suggestion
@description('Optional. how long this blob can be restored. It should be less than DeleteRetentionPolicy days.')
```

File: modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep (hunk 7 above)
Comment (ahmadabdalla) [reply modules/Microsoft.Storage/storageAccounts/blobServices/deploy.bicep:37]: Thanks. Added metadata, removed information from description and generated readme after applying suggestions

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**